### DISCUSSION TITLE
Is it possible to add LDAP authentication for nginx-s3-gateway?

### DISCUSSION THREAD

**Discussion by dmrlx** [nginxinc/nginx-s3-gateway#187] [Q&A]
For example, I have a configured nginx-s3-gateway and an AD containing users and I want to make it so that when trying to download a file through the S3 gateway the user must authenticate. 

I am new to nginx, can you tell me which way to "dig"?

**Comment by 4141done** [comment]
Hello! So I'm not an expert with LDAP, but it's possible that you could adapt the solution presented here:
* Blog post: https://www.nginx.com/blog/nginx-plus-authenticate-users/
* Github repo: https://github.com/nginxinc/nginx-ldap-auth

It talks a lot about NGINX Plus but the solution is applicable to the normal NGINX as well. It leverages the [auth_request](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html) directive and an external daemon.

I can't necessarily provide you with a full adaptation, but you should be able to add the `auth_request` directive to the configuration of the s3 gateway in [this area](https://github.com/nginxinc/nginx-s3-gateway/blob/a95a6f3479ff7fd128fa9e2a42b859887a457394/common/etc/nginx/templates/default.conf.template#L77) along with the other elements of the solution above.

**Reply by dmrlx** [reply]
Hi @4141done,


Thanks for your answer. I approached the problem using an alternative solution. To serve multiple buckets, I set up multiple Docker containers running nginx-s3-gateway. Additionally, I have an "external" nginx instance running on the system, which serves two key functions: directing requests to the appropriate ports (and, accordingly, to the right container) based on the address and handling LDAP authentication.

**Reply by 4141done** [reply]
Thank you for sharing your solution so that others (including myself 😅 ) can benefit.  I think that sounds like a great setup and gives you a lot more flexibility managing access.  Thanks for bringing up the use case
